### PR TITLE
Proof contract unit tests

### LINKED ISSUE #143: Unit tests for Proof contract

### PR DESCRIPTION
**Pull request by kreloaded**
Fixes #143 
- Created test cases for Proof contract
   - Proof::setupProof
   - Proof::proveStorageAccount
   - Proof::proveStorageExistence
   - Proof::storagePath

### CODE REVIEW

File: contracts/anchor/AnchorI.sol [hunk 1]
```diff
@@ -35,7 +35,6 @@ interface AnchorI {
      */
     function getStateRoot(uint256 _blockHeight)
         external
-        view
```
Comment (0xsarvesh) [inline contracts/anchor/AnchorI.sol]: Why this is removed?

File: contracts/test/anchor/SpyAnchor.sol [hunk 2]
```diff
@@ -30,12 +29,16 @@ contract SpyAnchor is MasterCopyNonUpgradable, AnchorI{
         require(false, 'SpyAnchor::getLatestStateRootBlockHeight should not be called.');
     }
 
-    function getStateRoot(uint256)
+    /**
+    * @dev _blockheight is not assigned to spyBlockHeight
```
Comment (0xsarvesh) [inline contracts/test/anchor/SpyAnchor.sol]: please fix the indentation.

File: contracts/test/anchor/SpyAnchor.sol [hunk 3]
```diff
@@ -30,12 +29,16 @@ contract SpyAnchor is MasterCopyNonUpgradable, AnchorI{
...
+    *       variable because it is view in it's function signature
+    */
+
+    function getStateRoot(uint256 _blockHeight)
```
Comment (0xsarvesh) [inline contracts/test/anchor/SpyAnchor.sol]:
```suggestion
    function getStateRoot(uint256)
```

File: contracts/test/anchor/SpyAnchor.sol [hunk 4]
```diff
@@ -30,12 +29,16 @@ contract SpyAnchor is MasterCopyNonUpgradable, AnchorI{
...
         external
-        view
```
Comment (0xsarvesh) [inline contracts/test/anchor/SpyAnchor.sol]: Why view is removed here?

File: contracts/test/message_bus/ProofDouble.sol [hunk 5]
```diff
@@ -0,0 +1,108 @@
+pragma solidity >=0.5.0 <0.6.0;
+
+// Copyright 2019 OpenST Ltd.
```
Comment (0xsarvesh) [inline contracts/test/message_bus/ProofDouble.sol]:
```suggestion
// Copyright 2020 OpenST Ltd.
```

File: contracts/test/message_bus/ProofDouble.sol [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import "../../message-bus/Proof.sol";
+import "../../message-bus/StateRootI.sol";
+import "../../lib/CircularBufferUint.sol";
```
Comment (0xsarvesh) [inline contracts/test/message_bus/ProofDouble.sol]: is it used somewhere?  Can't find the usage.

File: contracts/test/message_bus/ProofDouble.sol [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+contract ProofDouble is Proof {
+
+    /**
```
Comment (0xsarvesh) [inline contracts/test/message_bus/ProofDouble.sol:26]: Add `external` function section header.

File: test/message_bus/Proof/prove_storage_account.js [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+const { AccountProvider } = require('../../test_lib/utils.js');
+const ProveStorageAccount = require('./prove_storage_account_proof.json');
+
+contract('Proof::proveStorageAccount', async (accounts) => {
+  const accountProvider = new AccountProvider(accounts);
+  let proof;
+  let spyAnchor;
+  let setupParams;
+  let consensus;
```
Comment (0xsarvesh) [inline test/message_bus/Proof/prove_storage_account.js]: This variable can be local to `beforeEach`, not used outside the scope.

File: test/message_bus/Proof/prove_storage_account.js [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+          ProveStorageAccount.rlpAccountNode,
+          ProveStorageAccount.rlpParentNodes,
+        ),
+        'State root must not be zero.',
+      );
+    });
+  });
+
+
+  contract('Positive Tests', async () => {
```
Comment (0xsarvesh) [inline test/message_bus/Proof/prove_storage_account.js:115]: add more tests: 
1.  should store the latest storage root and delete oldest storage root when the circular buffer is full.
2. should not prove storage account if already proven.

File: test/message_bus/Proof/prove_storage_account.js [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+      );
+
+      ExpectedStorageRoot = await proof.storageRoots.call(ProveStorageAccount.blockNumber);
+
+      CalculatedStorageRoot = ProveStorageAccount.storageHash;
+
+      assert.strictEqual(
+        ExpectedStorageRoot,
+        CalculatedStorageRoot,
+        'Storage root/hash generated must match with storage root calculated.',
```
Comment (0xsarvesh) [inline test/message_bus/Proof/prove_storage_account.js]:
```suggestion
        'Storage root hash generated must match with storage root calculated.',
```

File: test/message_bus/Proof/prove_storage_existence.js [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+      setupParams.stateRootProvider,
+      setupParams.maxStorageRootItems,
+    );
+
+    await proof.setStorageRootDouble(
+      ProveStorageExistence.blockNumber,
+      ProveStorageExistence.storageHash,
+    );
+
+    path = await proof.storagePathDouble(
```
Comment (0xsarvesh) [inline test/message_bus/Proof/prove_storage_existence.js]: it's a constant value.  In my opinion, the test shouldn't depend on the implementation of other methods.

File: test/message_bus/Proof/prove_storage_existence.js [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+        ),
+        'Merkle proof verification failed.',
+      );
+    });
+  });
+
+
+  contract('Positive Tests', async () => {
+    it('should pass when existence account proof matches', async () => {
+      const value = web3.utils.soliditySha3(true);
```
Comment (0xsarvesh) [inline test/message_bus/Proof/prove_storage_existence.js]: should `value` be moved to JSON?

File: test/message_bus/Proof/prove_storage_existence_proof.json [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+  "blockNumber": 25150,
+  "stateRoot": "0xd4e7b2b2dcc46cb2811596bbb2afa30734e5156b1e3b29bbb2c9253894ae8626",
+  "inboundChannelIdentifier": "0x9cf5aac02a270245805fdcbf9f8c028225dafea560e9d175b4f9efa8462a7ff5",
+  "outboundChannelIdentifier": "0x1e224f8547dc5a23445814bdb16a8e3434b1f598b9b8af72c8e690d30f1c2776",
+  "kernelHash": "0xad7c5bef027816a800da1736444fb58a807ef4c9603b7848673f7e3a68eb14a5",
+  "sender": "0x40Ebe1ce3428b9CbAddBc20aF27F1c4780b7D49F",
+  "nonce": 0,
+  "metachainId": "0xc89efdaa54c0f20c7adf612882df0950f5a951637e0307cdcb4c672f298b8bc6"
```
Comment (0xsarvesh) [inline test/message_bus/Proof/prove_storage_existence_proof.json]: is this relevant in the context of this test? Can you also remove other params?

File: test/message_bus/Proof/setup_proof.js [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+
+'use strict';
+
+const Proof = artifacts.require('ProofDouble');
+const BN = require('bn.js');
+
+const { AccountProvider } = require('../../test_lib/utils.js');
+const Utils = require('../../test_lib/utils.js');
+
+contract('Proof::proveStorageAccount', (accounts) => {
```
Comment (0xsarvesh) [inline test/message_bus/Proof/setup_proof.js]:
```suggestion
contract('Proof::setupProof', (accounts) => {
```

File: test/message_bus/Proof/setup_proof.js [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+
+      const storageAccount = await proof.storageAccount.call();
+      assert.strictEqual(
+        setupParams.storageAccount,
+        storageAccount,
+        'Storage account address must match with the storage adress provided.',
+      );
+
+      const stateRootProvider = await proof.stateRootProvider.call();
+      assert.strictEqual(
```
Comment (0xsarvesh) [inline test/message_bus/Proof/setup_proof.js:107]: Please add an assertion for `encodedAccountPath`

File: test/message_bus/Proof/storage_path.js [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+
+const Proof = artifacts.require('ProofDouble');
+
+const BN = require('bn.js');
+const web3 = require('../../test_lib/web3.js');
+
+const { AccountProvider } = require('../../test_lib/utils.js');
+const ProveStorageExistence = require('./prove_storage_existence_proof.json');
+
+contract('Proof::proveStorageExistence', async (accounts) => {
```
Comment (0xsarvesh) [inline test/message_bus/Proof/storage_path.js]:
```suggestion
contract('Proof::storagePath', async (accounts) => {
```

**Review by 0xsarvesh — CHANGES_REQUESTED** [review]
Thanks for the changes, nice work 👍  
There are some feedbacks and comments inline.

File: contracts/test/message_bus/ProofDouble.sol [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+     * @notice It is used to test Proof::setupProof
+     */
+    function setupProofDouble(
```
Comment (0xsarvesh) [inline contracts/test/message_bus/ProofDouble.sol]:
```suggestion
    function setupProofExternal(
```

Can you change, suffix to `External` for other functions as well.

Comment (kreloaded) [reply contracts/test/message_bus/ProofDouble.sol]: Done.

File: test/message_bus/Proof/storage_path.js (hunk 16 above)
Comment (kreloaded) [reply test/message_bus/Proof/storage_path.js]: Done.

File: test/message_bus/Proof/setup_proof.js (hunk 14 above)
Comment (kreloaded) [reply test/message_bus/Proof/setup_proof.js]: Done.

File: test/message_bus/Proof/prove_storage_existence_proof.json (hunk 13 above)
Comment (kreloaded) [reply test/message_bus/Proof/prove_storage_existence_proof.json]: Done. Removed unnecessary params

File: test/message_bus/Proof/prove_storage_account.js (hunk 10 above)
Comment (kreloaded) [reply test/message_bus/Proof/prove_storage_account.js]: Done.

File: contracts/test/message_bus/ProofDouble.sol (hunk 6 above)
Comment (kreloaded) [reply contracts/test/message_bus/ProofDouble.sol]: Yes. It doesn't make any sense. Removed

File: contracts/test/message_bus/ProofDouble.sol (hunk 5 above)
Comment (kreloaded) [reply contracts/test/message_bus/ProofDouble.sol]: Corrected.

File: contracts/test/anchor/SpyAnchor.sol (hunk 2 above)
Comment (kreloaded) [reply contracts/test/anchor/SpyAnchor.sol]: Done.

File: contracts/test/anchor/SpyAnchor.sol (hunk 3 above)
Comment (kreloaded) [reply contracts/test/anchor/SpyAnchor.sol]: Removed.

File: contracts/test/message_bus/ProofDouble.sol (hunk 7 above)
Comment (kreloaded) [reply contracts/test/message_bus/ProofDouble.sol:26]: Added `external` function section header

File: test/message_bus/Proof/prove_storage_account.js (hunk 8 above)
Comment (kreloaded) [reply test/message_bus/Proof/prove_storage_account.js]: Corrected.

File: test/message_bus/Proof/prove_storage_existence.js (hunk 11 above)
Comment (kreloaded) [reply test/message_bus/Proof/prove_storage_existence.js]: Yes. Calculated separately.

File: test/message_bus/Proof/prove_storage_existence.js (hunk 12 above)
Comment (kreloaded) [reply test/message_bus/Proof/prove_storage_existence.js]: Yes it should be. `value` is now moved to JSON.

File: test/message_bus/Proof/setup_proof.js (hunk 15 above)
Comment (kreloaded) [reply test/message_bus/Proof/setup_proof.js:107]: Done.

File: contracts/anchor/AnchorI.sol (hunk 1 above)
Comment (kreloaded) [reply contracts/anchor/AnchorI.sol]: In the SpyAnchor::getStateRoot the blockHeight has to be updated but later removed it.
Fixed now.

File: contracts/test/anchor/SpyAnchor.sol (hunk 4 above)
Comment (kreloaded) [reply contracts/test/anchor/SpyAnchor.sol]: It is fixed now.

File: contracts/test/anchor/SpyAnchor.sol [hunk 18]
```diff
@@ -41,7 +41,7 @@ contract SpyAnchor is MasterCopyNonUpgradable, AnchorI{
         view
         returns (uint256)
     {
-        require(false, "SpyAnchor::getLatestStateRootBlockHeight should not be called.");
+        require(false, 'SpyAnchor::getLatestStateRootBlockHeight should not be called.');
```
Comment (0xsarvesh) [inline contracts/test/anchor/SpyAnchor.sol]:
```suggestion
        require(false, "SpyAnchor::getLatestStateRootBlockHeight should not be called.");
```

File: test/message_bus/Proof/prove_storage_account.js [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+        CalculatedStorageRoot,
+        'Storage root hash generated must match with storage root calculated.',
+      );
+    });
+
+    it('should pass when it is not proving already proved storage account', async () => {
+      const gasEstimatedNewStorageAccount = await proof.proveStorageAccountExternal.estimateGas(
+        new BN(ProveStorageAccount.blockNumber),
+        ProveStorageAccount.rlpAccountNode,
+        ProveStorageAccount.rlpParentNodes
```
Comment (0xsarvesh) [inline test/message_bus/Proof/prove_storage_account.js]:
```suggestion
        ProveStorageAccount.rlpParentNodes,
```

File: test/message_bus/Proof/prove_storage_account.js [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+      );
+
+      // setting storage account
+      await proof.setStorageRootExternal(
+        ProveStorageAccount.blockNumber,
+        ProveStorageAccount.storageHash
```
Comment (0xsarvesh) [inline test/message_bus/Proof/prove_storage_account.js]:
```suggestion
        ProveStorageAccount.storageHash,
```

File: test/message_bus/Proof/prove_storage_account.js [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+      )
+      const gasEstimatedExistingStorageAccount = await proof.proveStorageAccountExternal.estimateGas(
+        new BN(ProveStorageAccount.blockNumber),
+        ProveStorageAccount.rlpAccountNode,
+        ProveStorageAccount.rlpParentNodes
```
Comment (0xsarvesh) [inline test/message_bus/Proof/prove_storage_account.js]:
```suggestion
        ProveStorageAccount.rlpParentNodes,
```

File: test/message_bus/Proof/prove_storage_account.js [hunk 22]
```diff
@@ -0,0 +1,195 @@
...
+      );
+
+      // setting storage account
+      await proof.setStorageRootExternal(
+        ProveStorageAccount.blockNumber,
+        ProveStorageAccount.storageHash
+      )
```
Comment (0xsarvesh) [inline test/message_bus/Proof/prove_storage_account.js]:
```suggestion
      );
```

File: test/message_bus/Proof/prove_storage_account.js [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+      CalculatedStorageRoot = ProveStorageAccount.storageHash;
+
+      assert.strictEqual(
+        ExpectedStorageRoot,
+        CalculatedStorageRoot,
+        'Storage root hash generated must match with storage root calculated.',
+      );
+    });
+
+    it('should pass when it is not proving already proved storage account', async () => {
```
Comment (0xsarvesh) [inline test/message_bus/Proof/prove_storage_account.js]:
```suggestion
    it('should skip the proof verification if already proven for same block number', async () => {
```

File: test/message_bus/Proof/prove_storage_existence.js [hunk 24]
```diff
@@ -0,0 +1,105 @@
...
+      setupParams.stateRootProvider,
+      setupParams.maxStorageRootItems,
+    );
+
+    await proof.setStorageRootExternal(
+      ProveStorageExistence.blockNumber,
+      ProveStorageExistence.storageHash,
+    );
+
+    path = web3.utils.padLeft(ProveStorageExistence.messageHash,64)
```
Comment (0xsarvesh) [inline test/message_bus/Proof/prove_storage_existence.js]:
```suggestion
    path = web3.utils.padLeft(ProveStorageExistence.messageHash, 64);
```

File: test/message_bus/Proof/prove_storage_existence.js [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+    path = path + web3.utils.padLeft("1",64)
```
Comment (0xsarvesh) [inline test/message_bus/Proof/prove_storage_existence.js]:
```suggestion
    path += web3.utils.padLeft('1', 64);
```

File: test/message_bus/Proof/prove_storage_existence.js [hunk 26]
```diff
@@ -0,0 +1,105 @@
...
+          ProveStorageExistence.serializedStorageProof,
+        ),
+        'Merkle proof verification failed.',
+      );
+    });
+  });
+
+
+  contract('Positive Tests', async () => {
+    it('should pass when existence account proof matches', async () => {
```
Comment (0xsarvesh) [inline test/message_bus/Proof/prove_storage_existence.js]: Please rephrase this: it's not clear with the statement, what is tested here.

File: test/message_bus/Proof/setup_proof.js [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+    setupParams = {
+      storageAccount: accountProvider.get(),
+      stateRootProvider: accountProvider.get(),
+      maxStorageRootItems: new BN(100),
+    };
+  });
+
+
+  contract('Negative Tests', async () => {
+    it('should fail when stateRootProvider address is empty', async () => {
```
Comment (0xsarvesh) [inline test/message_bus/Proof/setup_proof.js]: I think statement and test is not aligned.

**Review by 0xsarvesh — CHANGES_REQUESTED** [review]
Nice progress 👍 

Some minor changes required.

File: contracts/test/anchor/SpyAnchor.sol (hunk 18 above)
Comment (kreloaded) [reply contracts/test/anchor/SpyAnchor.sol]: Done.

File: test/message_bus/Proof/prove_storage_account.js (hunk 19 above)
Comment (kreloaded) [reply test/message_bus/Proof/prove_storage_account.js]: Done.

File: test/message_bus/Proof/prove_storage_account.js (hunk 21 above)
Comment (kreloaded) [reply test/message_bus/Proof/prove_storage_account.js]: Done.

File: test/message_bus/Proof/prove_storage_account.js (hunk 22 above)
Comment (kreloaded) [reply test/message_bus/Proof/prove_storage_account.js]: Done.

File: test/message_bus/Proof/prove_storage_existence.js (hunk 24 above)
Comment (kreloaded) [reply test/message_bus/Proof/prove_storage_existence.js]: Corrected.

File: test/message_bus/Proof/prove_storage_existence.js (hunk 25 above)
Comment (kreloaded) [reply test/message_bus/Proof/prove_storage_existence.js]: Corrected.

File: test/message_bus/Proof/prove_storage_existence.js (hunk 26 above)
Comment (kreloaded) [reply test/message_bus/Proof/prove_storage_existence.js]: Yes. Corrected.

File: test/message_bus/Proof/setup_proof.js (hunk 27 above)
Comment (kreloaded) [reply test/message_bus/Proof/setup_proof.js]: Corrected.

**Review by 0xsarvesh — APPROVED** [review]
LGTM 👍

**Pull request merged**